### PR TITLE
Add a guide to writing an Access Request plugin

### LINKED ISSUE #22227: Add instructions for building a custom Access Request plugin
## Applies To

`docs/pages/api/*`

## Details
See #19716

Building Access Request plugins is a popular use case for our API. Let's add a how-to guide that walks a user through how to write their own Access Request plugin, sending Access Request notifications to a generic HTTP REST/JSON API endpoint.

## Related Issues
- #19716

### PR DESCRIPTION
**Pull request by ptgott**
Closes #22227

This guide uses a minimal working example of an Access Request plugin to demonstrate some of the key libraries that Teleport makes available to plugin developers. The guide is based on a 270-line plugin that manages Access Requests using Google Sheets.

### CODE REVIEW

File: docs/pages/api/access-plugin.mdx [hunk 1]
```diff
@@ -0,0 +1,738 @@
+---
```
Comment (hugoShaka) [inline docs/pages/api/access-plugin.mdx:1]: This guide provides a great minimal example of how to consume incoming events :). It relies on the fact you can walk through all rows in the spreadsheet to find the one to update, making the process stateless.

Real-world plugins almost always have to introduce a stateful component (the message ID the channel was sent to, whether the event was already processed, ...). This allows to increase reliability and run replicated plugins. A step by step how-to about those topics would not be practical nor useful, but it might be useful to mention the teleport pluginData API in the `NextSteps` section. Essentially telling that it exists, uses a compare-and-swap approach to deal with concurrency, and that it allows to process events in a stateful way.

We could also mention that we strongly discourage handling the approval at the plugin level (e.g. with a message answer) because it could lead to severe security issues where an unauthorized person can respond to the plugin and escalate privileges.

File: docs/pages/api/access-plugin.mdx [hunk 2]
```diff
@@ -0,0 +1,738 @@
...
+title: How to Build an Access Request Plugin
+description: Manage Access Requests using custom workflows by using the Teleport API
```
Comment (zmb3) [inline docs/pages/api/access-plugin.mdx:3]:
```suggestion
description: Manage Access Requests using custom workflows with the Teleport API
```

To avoid saying "using" twice

File: docs/pages/api/access-plugin.mdx [hunk 3]
```diff
@@ -0,0 +1,738 @@
...
+The plugin we will build in this guide is intended as a learning tool. **Do not
+connect it to your production Teleport cluster.**  Use a demo cluster instead.
+
+</Notice>
+
+## Prerequisites
+
+(!docs/pages/includes/commercial-prereqs-tabs.mdx!)
+
+- Go installed on your workstation. See the [Go download
```
Comment (zmb3) [inline docs/pages/api/access-plugin.mdx:35]: Worth mentioning the minimum go version? (as declared in api/go.mod)

File: docs/pages/api/access-plugin.mdx [hunk 4]
```diff
@@ -0,0 +1,738 @@
...
+
+Whether creating a new row of the spreadsheet or updating an existing one, we
+need a way to extract data from an Access Reqeust in order to provide it to
+Google Sheets. Below the `googleSheetsPlugin` type declaration, add the
+following function:
+
+```go
+func (g *googleSheetsPlugin) makeRowData(ar types.AccessRequest) *sheets.RowData {
+
+	t := ar.GetCreationTime().String()
```
Comment (zmb3) [inline docs/pages/api/access-plugin.mdx:302]: I find this hard to follow, since the variables have non-descriptive names and are used a dozen or so lines after they are declared.

How about a helper:

```
func stringPtr(s string) *string { return &s 
```

Then you don't need these variables here and you inline them below.

```
UserEnteredValue: &sheets.ExtendedValue{
    StringValue: stringPtr(ar.GetCreationTime().String()),
},
```

File: docs/pages/api/access-plugin.mdx [hunk 5]
```diff
@@ -0,0 +1,738 @@
...
+	u := ar.GetUser()
+	s := strings.Join(ar.GetRoles(), ",")
+	i := ar.GetName()
+	a, ok := requestStates[ar.GetState()]
+	l := fmt.Sprintf(
+		"=HYPERLINK(\"%v\", \"%v\")",
```
Comment (zmb3) [inline docs/pages/api/access-plugin.mdx:308]:
```suggestion
		`=HYPERLINK("%v", "%v")`,
```

File: docs/pages/api/access-plugin.mdx [hunk 6]
```diff
@@ -0,0 +1,738 @@
...
+		"https://"+proxyAddr+"/web/requests/"+ar.GetName(),
+		"View Access Request",
+	)
+
+	// Could not find a state, but this is still a valid Access Request
+	if !ok {
```
Comment (zmb3) [inline docs/pages/api/access-plugin.mdx:314]: Move this right after the `a, ok :=` assignment. Having code in between makes this harder to follow.

File: docs/pages/api/access-plugin.mdx [hunk 7]
```diff
@@ -0,0 +1,738 @@
...
+					Rows: []*sheets.RowData{
+						row,
+					},
+				},
+			},
+		},
+	}
+
+	resp, err := g.sheetsClient.BatchUpdate(spreadSheetID, &req).Do()
+
```
Comment (zmb3) [inline docs/pages/api/access-plugin.mdx:399]: Whitespace is a great way to separate logical pieces of code. In this case, though, the error handling should be considered part of the same task.
```suggestion
```

File: docs/pages/api/access-plugin.mdx [hunk 8]
```diff
@@ -0,0 +1,738 @@
...
+can retry it later.
+
+### Update a row
+
+The code for updating a row is similar to the code for creating a new row. Add
+this function below `createRow`:
+
+```go
+func (g *googleSheetsPlugin) updateRow(ar types.AccessRequest, rowNum int64) error {
+
```
Comment (zmb3) [inline docs/pages/api/access-plugin.mdx:438]:
```suggestion
```

File: docs/pages/api/access-plugin.mdx [hunk 9]
```diff
@@ -0,0 +1,738 @@
...
+					},
+					Rows: []*sheets.RowData{
+						row,
+					},
+				},
+			},
+		},
+	}
+	resp, err := g.sheetsClient.BatchUpdate(spreadSheetID, &req).Do()
+
```
Comment (zmb3) [inline docs/pages/api/access-plugin.mdx:458]:
```suggestion
```

File: docs/pages/api/access-plugin.mdx [hunk 10]
```diff
@@ -0,0 +1,738 @@
...
+Below `updateRow`, add this function:
+
+```go
+func (g *googleSheetsPlugin) updateSpreadsheet(ar types.AccessRequest) error {
+
+	s, err := g.sheetsClient.Get(spreadSheetID).IncludeGridData(true).Do()
+
+	if err != nil {
+		return trace.Wrap(err)
```
Comment (zmb3) [inline docs/pages/api/access-plugin.mdx:492]: Either use trace consistently, or don't use it. A reader unfamiliar with Go will be wondering why we're not consistent here.

File: docs/pages/api/access-plugin.mdx [hunk 11]
```diff
@@ -0,0 +1,738 @@
...
+function takes the number of a row within the spreadsheet to update and sends a
+request to update that row with information from the provided Access Request.
+
+### Determine where to update the spreadsheet
+
+Below `updateRow`, add this function:
+
+```go
+func (g *googleSheetsPlugin) updateSpreadsheet(ar types.AccessRequest) error {
+
```
Comment (zmb3) [inline docs/pages/api/access-plugin.mdx:488]:
```suggestion
```

File: docs/pages/api/access-plugin.mdx [hunk 12]
```diff
@@ -0,0 +1,738 @@
...
+	s, err := g.sheetsClient.Get(spreadSheetID).IncludeGridData(true).Do()
+
```
Comment (zmb3) [inline docs/pages/api/access-plugin.mdx:490]:
```suggestion
```

File: docs/pages/api/access-plugin.mdx [hunk 13]
```diff
@@ -0,0 +1,738 @@
...
+func (g *googleSheetsPlugin) handleEvent(ctx context.Context, event types.Event) error {
+
+	if event.Resource == nil {
+		return nil
+	}
+
+	r := event.Resource.(types.AccessRequest)
+
+	// Ignore Access Requests that are not Role Requests
+	if len(r.GetRoles()) == 0 {
```
Comment (zmb3) [inline docs/pages/api/access-plugin.mdx:539]: Don't resource access requests have role(s) too?

File: docs/pages/api/access-plugin.mdx (hunk 1 above)
Comment (ptgott) [reply docs/pages/api/access-plugin.mdx:1]: Thanks! I've added this information

File: docs/pages/api/access-plugin.mdx (hunk 4 above)
Comment (ptgott) [reply docs/pages/api/access-plugin.mdx:302]: Added for the one-line assignments. For the request state and link string, I used descriptive variable names instead.

File: docs/pages/api/access-plugin.mdx (hunk 13 above)
Comment (ptgott) [reply docs/pages/api/access-plugin.mdx:539]: This was a bad assumption on my part based on not being familiar enough with Resource Access Requests. I'll remove this.

**Comment by ptgott** [comment]
@alexfornuto This should be ready for a copy review

File: docs/pages/api/access-plugin.mdx [hunk 14]
```diff
@@ -0,0 +1,761 @@
...
+
+Click **Create and Continue**. When prompted to grant roles to the service
+account, click **Continue** again. We will create our service account without
+roles. Skip the step to grant users access to the service account, clicking
+**Done**.
+
+The console will take you to the **Service accounts** view. Click the name of
+the service account you just created, then click the **Keys** tab. Click **Add
+Key**, then **Create new key**. Leave the **Key type** as "JSON" and click
+**Create**.
```
Comment (alexfornuto) [inline docs/pages/api/access-plugin.mdx:115]: I suggest telling the reader to save the file as `credentials.json`, since that filename is expected in the code below.

File: docs/pages/api/access-plugin.mdx [hunk 15]
```diff
@@ -0,0 +1,761 @@
...
+</Admonition>
+
+## Step 7/7. Test your plugin
+
+Now that you have written your plugin, run it to forward Access Requests from
+your Teleport cluster to Google Sheets. Execute the following command from
+within your project directory:
+
+```code
+$ go run main.go
```
Comment (alexfornuto) [inline docs/pages/api/access-plugin.mdx:698]: I think we need a `go get` first. I did.

File: docs/pages/api/access-plugin.mdx [hunk 16]
```diff
@@ -0,0 +1,761 @@
...
+```
+
+Now that the plugin is running, create an Access Request:
+
+(!docs/pages/includes/plugins/create-request.mdx!)
+
+You should see the new Access Request in your spreadsheet with the `PENDING`
+state.
+
+Sign into the Teleport Web UI as your original user. In your spreadsheet,  click
```
Comment (alexfornuto) [inline docs/pages/api/access-plugin.mdx:708]: The first two sentences can be swapped, since you don't need to independently open the web UI and sign in before clicking the link; Teleport will prompt you to sign in and redirect from that link.

File: docs/pages/api/access-plugin.mdx [hunk 17]
```diff
@@ -0,0 +1,761 @@
...
+Access Request plugin uses the SSH credentials to connect to the Proxy Service,
+which establishes a reverse tunnel connection to the Auth Service. The plugin
+uses this reverse tunnel, along with your TLS credentials, to connect to the
+Auth Service's gRPC endpoint.
+
+You will refer to this file later when configuring the plugin.
+
+### Set up Role Access Requests
+
+In this guide, we will use our plugin to manage Role Access Requests. For this
```
[the remark this answers is not in the document]
Comment (ptgott) [reply docs/pages/api/access-plugin.mdx:185]: Not sure about this, since we don't use the plugin in this section.

**Review by alexfornuto — APPROVED** [review]

**Comment by ptgott** [comment]
@hugoShaka @zmb3 Just checking to see if you were planning to give this another review. Thanks!

**Review by hugoShaka — APPROVED** [review]

**Review by zmb3 — APPROVED** [review]

**Comment by stevenGravy** [comment]
@ptgott some misspellings here.  I'll post in here the `master` PR so you can apply them in the backports.

**Comment by stevenGravy** [comment]
Spell fixes in #23068
